### DISCUSSION TITLE
node v12 support for optional chaining is broken?

### DISCUSSION THREAD

**Discussion by geoidesic** [babel/babel#15247] [Q&A] [answered]
Hi there,

We've been merrily using babel's support for transpiling JS optional chaining syntax to node v12 in our v1.quasar (Vue v2) app for more than a year. Last week it stopped working and now we're seeing errors like this:

```
Module parse failed: Unexpected token (16:20)
File was processed with these loaders:
 * ./node_modules/@quasar/app/lib/webpack/loader.transform-quasar-imports.js
 * ./node_modules/babel-loader/lib/index.js
 * ./node_modules/@quasar/app/lib/webpack/loader.auto-import-client.js
 * ./node_modules/vue-loader/lib/index.js
You may need an additional loader to handle the result of these loaders.
|     },
|     bb() {
>       return this.b?.a;
|     }
|   }
```

We use a `babel.config.js` file:
```
/* eslint-env node */
// eslint-disable-next-line @typescript-eslint/no-var-requires
const fs = require("fs-extra");
let extend = undefined;

/**
 * The .babelrc file has been created to assist Jest for transpiling.
 * You should keep your application's babel rules in this file.
 */

if (fs.existsSync("./.babelrc")) {
  extend = "./.babelrc";
}

module.exports = {
  presets: ["@quasar/babel-preset-app"],
  extends: extend,
};
```

And the `package.json` looks like this:
```
{
  "name": "quasar-project",
  "version": "0.0.1",
  "description": "A Quasar Project",
  "productName": "Quasar App",
  "author": "",
  "private": true,
  "scripts": {
    "lint": "eslint --ext .js,.vue ./",
    "test": "echo \"No test specified\" && exit 0"
  },
  "dependencies": {
    "@babel/core": "^7.20.5",
    "@quasar/extras": "^1.0.0",
    "axios": "^0.21.1",
    "core-js": "^3.6.5",
    "quasar": "^1.0.0",
    "timezone-support": "1.8.1",
    "vue-i18n": "^8.0.0",
    "webpack": "^4.0.0"
  },
  "devDependencies": {
    "@babel/eslint-parser": "^7.13.14",
    "@quasar/app": "^2.0.0",
    "eslint": "^8.10.0",
    "eslint-config-prettier": "^8.1.0",
    "eslint-plugin-vue": "^8.0.0",
    "eslint-webpack-plugin": "^2.0.0",
    "prettier": "^2.5.1"
  },
  "browserslist": [
    "last 10 Chrome versions",
    "last 10 Firefox versions",
    "last 4 Edge versions",
    "last 7 Safari versions",
    "last 8 Android versions",
    "last 8 ChromeAndroid versions",
    "last 8 FirefoxAndroid versions",
    "last 10 iOS versions",
    "last 5 Opera versions"
  ],
  "engines": {
    "node": ">= 10.18.1",
    "npm": ">= 6.13.4",
    "yarn": ">= 1.21.1"
  }
}
```

Any ideas what might have gone wrong?

**Comment by JLHwung** [comment]
I guess after you updated `caniuse-lite`, the browsers you specified support optional chaining so Babel does not transform that at all.

- If your app will run at Node.js 10, please add `node 10.18.1` to the `browserslist`.
- If your app does not run at Node but you are transpiling optional chaining because of the error "You may need an additional loader to handle the result of these loaders.", please update webpack to v5.

**Comment by geoidesic** [comment]
I tried "node 10.18.1" and "node 12.22.12" in the list but it yields an error:
```

 App · ⚠️  BrowserslistError: Unknown version 12.22.12 of Node.js
 App · ⚠️  Please revise /package.json > browserslist -- invalid entry!

```

**Comment by geoidesic** [comment]
The app is a vueJS app running node 12 as a compiler.

**Comment by geoidesic** [comment, marked as answer]
"maintained node versions" did the trick!

**Reply by JLHwung** [reply]
"maintained node versions" should not include Node 12 now. It went end of life in Apr 2022.
> BrowserslistError: Unknown version 12.22.12 of Node.js

You are probably using out-of-dated `browserslist` and `node-releases`, please update both.

**Comment by geoidesic** [comment]
@JLHwung I googled `how to update node-releases` and drew a blank. I found this but there's no documentation so I'm not sure if it's related: https://github.com/chicoxyzzy/node-releases. Any clues?

I'm now trying this with node v16 and I'm still getting the same problem, optional chaining syntax is not recognised by the compiler, even with "maintained node versions" in the browser list. Please advise.

**Reply by JLHwung** [reply]
You browserslist targets already support optional chaining. So `preset-env` does not enable the optional chaining plugin. However, webpack 4 does not support optional chaining, so please upgrade to webpack 5.
